### PR TITLE
[Jetsurvey] Add permissions

### PR DESCRIPTION
**Pull request by manuelvicnt**
Screenshot 2021-06-15 at 18 23 38
Screenshot 2021-06-15 at 18 58 07
Screenshot 2021-06-15 at 18 58 14
Screenshot 2021-06-15 at 18 58 25

### CODE REVIEW

File: Jetsurvey/app/src/main/AndroidManifest.xml [hunk 1]
```diff
@@ -17,6 +17,7 @@
 <manifest xmlns:android="http://schemas.android.com/apk/res/android"
     package="com.example.compose.jetsurvey">
 
+    <uses-permission android:name="android.permission.CAMERA"/>
```
Comment (yrezgui) [inline Jetsurvey/app/src/main/AndroidManifest.xml:20]: I don't think you need this permission. You can trigger the intent without requesting it

File: Jetsurvey/app/src/main/java/com/example/compose/jetsurvey/survey/PhotoUriManager.kt [hunk 2]
```diff
@@ -26,7 +27,11 @@ import android.provider.MediaStore
 class PhotoUriManager(private val appContext: Context) {
 
     private val photoCollection by lazy {
-        MediaStore.Images.Media.getContentUri(MediaStore.VOLUME_EXTERNAL_PRIMARY)
+        if (Build.VERSION.SDK_INT > 28) {
```
Comment (yrezgui) [inline Jetsurvey/app/src/main/java/com/example/compose/jetsurvey/survey/PhotoUriManager.kt:30]: Use `Build.VERSION_CODES.P`

Comment (chrisbanes) [reply Jetsurvey/app/src/main/java/com/example/compose/jetsurvey/survey/PhotoUriManager.kt:30]: Nah, the preference is not to do that now, and to just hardcover the number

File: Jetsurvey/app/src/main/AndroidManifest.xml (hunk 1 above)
Comment (chrisbanes) [reply Jetsurvey/app/src/main/AndroidManifest.xml:20]: Pretty sure this is still required, but I'm not 100% sure

Comment (yrezgui) [reply Jetsurvey/app/src/main/AndroidManifest.xml:20]: I checked and it's not useful: https://developer.android.com/training/permissions/evaluating#take-photo

Comment (chrisbanes) [reply Jetsurvey/app/src/main/AndroidManifest.xml:20]: That only works if you use the use the intent to open the Camera app. This PR is all about requesting permissions, and this line is needed to demonstrate that AFAIK.

Comment (manuelvicnt) [reply Jetsurvey/app/src/main/AndroidManifest.xml:20]: Yacine is right, this question uses the `TakePicture` activity result contract that does NOT require the camera permission. I thought it was needed (well, at least, I tested the flow with more permissions :D haha)

The `WRITE_EXTERNAL_STORAGE` is still required for APIs lower than 29 so we'd be showcasing permissions anyway.

File: Jetsurvey/app/src/main/java/com/example/compose/jetsurvey/survey/Survey.kt [hunk 3]
```diff
@@ -34,7 +34,9 @@ data class Question(
     val id: Int,
     @StringRes val questionText: Int,
     val answer: PossibleAnswer,
-    @StringRes val description: Int? = null
+    @StringRes val description: Int? = null,
+    val permissionsRequired: List<String>? = null,
```
Comment (florina-muntenescu) [inline Jetsurvey/app/src/main/java/com/example/compose/jetsurvey/survey/Survey.kt:38]: How about initialising this with `emptyList()` and making it non-null?
Nullability doesn't bring anything to empty list.

File: Jetsurvey/app/src/main/java/com/example/compose/jetsurvey/survey/SurveyQuestions.kt [hunk 4]
```diff
@@ -69,12 +72,121 @@ import androidx.compose.ui.unit.dp
...
+    question: Question,
+    answer: Answer<*>?,
+    shouldAskPermissions: Boolean,
+    onAnswer: (Answer<*>) -> Unit,
+    onAction: (Int, SurveyActionType) -> Unit,
+    onDoNotAskForPermissions: () -> Unit,
+    openSettings: () -> Unit,
+    modifier: Modifier = Modifier
+) {
+    if (question.permissionsRequired == null || question.permissionsRequired.isEmpty()) {
```
Comment (florina-muntenescu) [inline Jetsurvey/app/src/main/java/com/example/compose/jetsurvey/survey/SurveyQuestions.kt:91]: see comment above about making `permissionsRequired` nonnull

File: Jetsurvey/app/src/main/java/com/example/compose/jetsurvey/survey/SurveyQuestions.kt [hunk 5]
```diff
@@ -69,12 +72,121 @@ import androidx.compose.ui.unit.dp
...
+            // The user denied some permissions but a rationale should be shown
+            multiplePermissionsState.shouldShowRationale -> {
+                if (!shouldAskPermissions) {
+                    PermissionsDenied(
+                        question.questionText,
+                        openSettings,
+                        permissionsContentModifier
+                    )
+                } else {
+                    Column(permissionsContentModifier) {
```
Comment (florina-muntenescu) [inline Jetsurvey/app/src/main/java/com/example/compose/jetsurvey/survey/SurveyQuestions.kt:116]: can we wrap this into a composable to make it easier to understand what you're displaying here?

**Review by florina-muntenescu — CHANGES_REQUESTED** [review]

File: Jetsurvey/app/src/main/java/com/example/compose/jetsurvey/survey/Survey.kt (hunk 3 above)
Comment (manuelvicnt) [reply Jetsurvey/app/src/main/java/com/example/compose/jetsurvey/survey/Survey.kt]: Done!

File: Jetsurvey/app/src/main/java/com/example/compose/jetsurvey/survey/SurveyQuestions.kt (hunk 5 above)
Comment (manuelvicnt) [reply Jetsurvey/app/src/main/java/com/example/compose/jetsurvey/survey/SurveyQuestions.kt]: Done

**Review by florina-muntenescu — APPROVED** [review]

File: Jetsurvey/app/src/main/java/com/example/compose/jetsurvey/survey/SurveyRepository.kt [hunk 6]
```diff
@@ -83,15 +83,21 @@ private val jetpackQuestions = mutableListOf(
             endText = R.string.strongly_like,
             neutralText = R.string.neutral
         )
-    )
+    ),
 ).apply {
-    if (Build.VERSION.SDK_INT > Build.VERSION_CODES.Q) {
-        // Add the camera feature only for devices 29+
+    // TODO: FIX! After taking the selfie, the picture doesn't appear in API 22 and lower.
```
Comment (chrisbanes) [inline Jetsurvey/app/src/main/java/com/example/compose/jetsurvey/survey/SurveyRepository.kt:88]:
```suggestion
    // FIXME: After taking the selfie, the picture doesn't appear in API 22 and lower.
```

**Review by chrisbanes — APPROVED** [review]

**Review by yrezgui — APPROVED** [review]

**Pull request merged**